text/3735-field-projections.md

```diff
@@ -0,0 +1,1789 @@
+types are generated for all[^3] fields of all structs. A field type always implements the
+`UnalignedField` trait:
+
+[^3]: Almost all, fields that don't have a size, are not included.
+
+```rust
+pub unsafe trait UnalignedField {
+    type Base: ?Sized;
+    type Type: ?Sized;
+    const OFFSET: usize;
+}
+```
+
+`Base` is set to the parent struct containing the field, `Type` is set to the type of the field
+itself and `OFFSET` is the offset in bytes as returned by `offset_of!`.
+
+
+With the above projections in place, users can write the following code:
+
+```rust
+struct Foo {
+    bar: i32,
+    baz: u32,
+}
+
+let x: &Foo = &Foo { bar: 42, baz: 43 };
+let y: &Foo = &Foo { bar: 24, baz: 25 };
+let d: DoubleRef<'_, Foo> = DoubleRef::new(x, y);
+let bars: DoubleRef<'_, i32> = d->bar;
+assert_eq!((42, 24), bars.read());
+```
+
+#### Containers
+
+The other kind of projections are that of *containers*, for example `UnsafeCell<T>` or
+`MaybeUninit<T>`. They are governed by the `#[projecting]` attribute.
+
+If we want to combine the two containers from the previous sections, we can do it in the following
+way:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Opaque<T> {
+    inner: UnsafeCell<MaybeUninit<T>>,
+}
+
+impl<T> Opaque<T> {
+    pub fn new(value: T) -> Self {
+        Self { inner: UnsafeCell::new(MaybeUninit::new(value)) }
+    }
+
+    pub fn uninit() -> Self {
+        Self { inner: UnsafeCell::new(MaybeUninit::uninit()) }
+    }
+
+    pub fn write(&mut self, value: T) {
+        self.inner.get_mut().write(value);
+    }
+}
+```
+
+Now `&Opaque<T>` can represent a reference that points at data from other languages, such as C.
+
+The `#[projecting]` attribute changes the way the field types are generated for this type. Instead
+of having a field type representing the `inner` field, this type will "project" through to the type
+`T`, inheriting all fields that `T` has. So if we consider the type `Foo` from above:
+
+```rust
+struct Foo {
+    bar: i32,
+    baz: u32,
+}
+```
+
+Then there are two field types for `Opaque<Foo>`:
+- one representing `bar` with `Base = Opaque<Foo>` and `Type = Opaque<i32>`
+- the other representing `baz` with `Base = Opaque<Foo>` and `Type = Opaque<u32>`
+
+So the both the base type and the type of the fields are wrapped with the container. Now users can
+write:
+
+```rust
+fn init(foo: &mut Opaque<Foo>) {
+    let bar: &mut Opaque<i32> = foo->bar;
+    bar.write(42);
+    foo->baz.write(24);
+}
+```
+
+
+## Impact of this Feature
+
+Overall this feature improves readability of code, because it replaces more complex to parse syntax
+with simpler syntax:
+- `&raw mut (*ptr).foo` is turned into `ptr->foo`
+- using `NonNull<T>` as a replacement for `*mut T` becomes a lot better when accessing fields,
+
+There is of course a cost associated with introducing a new operator along with the concept of field
+projections.
```

Comment:
One problem that I find very difficult to address is that of how hard this feature is to understand. Since I have been thinking about this feature a lot (and seeing its uses out in the wild for even longer) I feel like I have lost the ability to accurately gauge its difficulty. I also have no idea of any particular hurdles that people come across when learning about  field projections for the first time.

This also brings me to the question of how we teach this. I tried to write the guide-level explanation in a rust-book chapter-style, but I don't know if it even reaches that standard, or if we even would like to talk about field projections there.

Reply:
seems simple enough to me, then again I'm already familiar with C++ member pointers, so I probably won't be much help on teachability aspects...

Reply:
Is this the same concept as the c++ member pointers?